hudi-spark-datasource/hudi-spark3-extensions_2.12/src/main/scala/org/apache/hudi/execution/HudiSQLUtils.scala

```diff
@@ -0,0 +1,553 @@
+
+    // commit and post commit
+    commitAndPerformPostOperations(writeResult,
+      parameters,
+      client,
+      HoodieTableMetaClient.builder().setConf(spark.sparkContext.hadoopConfiguration).setBasePath(path.get)
+        .build().getTableConfig,
+      jsc,
+      TableInstantInfo(basePath, instantTime, commitActionType, WriteOperationType.UPSERT))
+  }
+
+  private def commitAndPerformPostOperations(
+      writeResult: HoodieWriteResult,
+      parameters: Map[String, String],
+      client: SparkRDDWriteClient[HoodieRecordPayload[Nothing]],
+      tableConfig: HoodieTableConfig,
+      jsc: JavaSparkContext,
+      tableInstantInfo: TableInstantInfo): Unit = {
+    val errorCount = writeResult.getWriteStatuses.rdd.filter(ws => ws.hasErrors).count()
+    if (errorCount == 0) {
+      log.info("No errors. Proceeding to commit the write.")
+      val metaMap = parameters.filter(kv => kv._1.startsWith(parameters(COMMIT_METADATA_KEYPREFIX_OPT_KEY)))
+      val commitSuccess =
+        client.commit(tableInstantInfo.instantTime, writeResult.getWriteStatuses,
+          org.apache.hudi.common.util.Option.of(new util.HashMap[String, String](mapAsJavaMap(metaMap))),
+          tableInstantInfo.commitActionType, writeResult.getPartitionToReplaceFileIds)
+      if (commitSuccess) {
+        log.info("Commit" + tableInstantInfo.instantTime + " successful!")
+      } else {
+        log.info("Commit " + tableInstantInfo.instantTime + " failed!")
+      }
+    } else {
+      log.info(s"${tableInstantInfo.operation} failed with $errorCount errors :")
+      if (log.isTraceEnabled) {
+        log.trace("Printing out the top 100 errors")
+        writeResult.getWriteStatuses.rdd.filter(ws => ws.hasErrors)
+          .take(100)
+          .foreach(ws => {
+            log.trace("Global error :", ws.getGlobalError)
+            if (ws.getErrors.size() > 0) {
+              ws.getErrors.asScala.foreach(kt => log.trace(s"Error for key: ${kt._1}", kt._2))
+            }
+          })
+      }
+    }
+  }
+
+  /**
+    * do merge directly, just use hoodie dataFrame api
+    */
+  private def upsertDirectly(
+      df: DataFrame,
+      options: java.util.Map[String, String],
+      writeOperationType: WriteOperationType = WriteOperationType.UPSERT,
+      mode: SaveMode,
+      savePath: String): Unit = {
+    val par = options.getOrDefault("par", "1500")
+    df.write.format("hudi")
+      .option(BULKINSERT_PARALLELISM, par)
+      .option(INSERT_PARALLELISM, par)
+      .option(UPSERT_PARALLELISM, par)
+      .option(DELETE_PARALLELISM, par)
+      .options(options)
+      .option(DataSourceWriteOptions.OPERATION_OPT_KEY, writeOperationType.value())
+      .mode(mode)
+      .save(savePath)
+  }
+
+  /**
+    * now just sync hudi table info to hive
+    * we use hive driver to do sync, maybe it's better to use spark
+    */
+  private def doSyncToHive(parameters: Map[String, String], basePath: Path, hadoopConf: Configuration): Unit = {
```

Comment:
since we also support sync to DLA or other metastore in a plugin way, would you please also consider?

Reply:
Sorry, I forgot to think about it， i will update code